Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -107,98 +125,141 @@
                                  1 -1 -1  1  1 -1 -1  1;
                                  1  1 -1 -1  1  1 -1 -1;
                                  1 -1  1 -1  1 -1  1 -1];
-		end    
-    end
+		end
+	end
     
-    if strcmp(xQ.Vendor,'Philips')
-        xQ.TimeEncodedMatrix = xQ.TimeEncodedMatrix * -1;
+    switch xQ.Vendor
+        case 'Siemens'
+            % This is where this code was initially based on, so for
+            % Siemens we don't do anything specific
+        case 'Philips'
+            xQ.TimeEncodedMatrix = xQ.TimeEncodedMatrix * -1;
+            % Philips uses an inverse matrix compared to Siemens
+        case 'GE'
+            warning('Time encoded ASL data detected for GE, but GE usually does the decoding in K-space on the scanner');
+        otherwise
+            warning(['Time encoded ASL data not yet tested for vendor ' xQ.vendor]);
     end
     
 end
 
-% Load time-series nifti
+
+%% 2. Load time-series nifti
 imEncoded = xASL_io_Nifti2Im(imPath); 
 
-% Calculate specific data sizes
-nDecodedTI = xQ.TimeEncodedMatrixSize-1;                   % Number of TIs is always MatrixSize -1
-nDecodedVolumes = xQ.NumberEchoTimes * nDecodedTI;
+% Prepare the general sequence parameters
+nEncodedVolumes = size(imEncoded, 4);     % Size of the raw data
+
+% Check that input quantification parameters are vectors with an equal length to the number of volumes
+if length(xQ.LabelingDuration)>1 && length(xQ.LabelingDuration) ~= nEncodedVolumes
```

Comment:
For these checks here, the length()>1 should be removed, because at the end of this function you apply the decoding indices which would crash for length()==1. If you want to allow length()==1 we should probably use the same general repmat() as I implemented at the start of xASL_module_ASL

Reply:
I agree. If length of LabelingDuration is 1, then we need to replicate the vector to the correct length. 

Same with the PLDs below.